### ISSUE TITLE
Infinite loop on print_grammar

### ISSUE THREAD

**Issue by markomanninen** [mspandit/sequitur-python#1]
Hi,

I made this observation when analyzing Revelation book content. Grammar class will encounter infinite loop when processing text. I found exact character from the data. Thus on attachment I have provided two texts, the first one will work well because additional space is added to the text string, while latter will dive into the infinite loop. **Would you take a look, what is really the cause of the problem?** I tested this with Python 2 and 3 environment (Jupyter notebooks). I also tested original text on sequitur.info. There it is working perfectly.

```
# works great
seq = open('revelation.txt', 'r').read()
# will give an infinite loop
#seq = open('revelation1.txt', 'r').read()
g = Grammar()
g.train_string(seq)
print (g.print_grammar())
```

[revelation1.txt](https://github.com/mspandit/sequitur-python/files/144218/revelation1.txt)
[revelation.txt](https://github.com/mspandit/sequitur-python/files/144219/revelation.txt)

I also found that by doing this according to Grammar class:

```
rule_set = [self.root_production]
[rule.print_rule(rule_set) for rule in rule_set]
```

will give an infinite loop, but when doing this:

```
rule_set = [self.root_production]
for rule in rule_set:
    print (rule.print_rule(rule_set))
```

rules will print well. So it seems that if rules are set back on list, strange behaviour begins.

**Comment by mspandit** [comment]
I have been able to reproduce the problem thanks to your message.

Unfortunately, the large size of the data set makes it difficult to pinpoint the defect.

It will take me some days to fix.

**Comment by mspandit** [comment]
Fixed in commit 0dad947.

**Comment by markomanninen** [comment]
Hi Milind,

well done, this fix works great! No more infinite loops.

I have made other minor changes/additions to the code, main functionality is to get the grammar in plain list rather than text formatted output with a new method: g.get_grammar().

Also module imports in Python 3 are different nowadays, so there are few things to consider on that side.

You may want to take a look to the changes in https://github.com/markomanninen/sequitur-python and related work on the other repository: https://github.com/markomanninen/pysequitur if you want to implement some of them.

Have a great time!
-Marko
